### PR TITLE
Persistence for blob-retention configuration

### PR DESCRIPTION
**Pull request by Shrekster**
This PR adds:

- an authenticated repository blob `kopia.retention` when the repository is created with blob-retention options.
- updates the change-password mechanism to rewrite the retention blob
- adds plumbing for retention options to be supplied to index and pack blobs puts

### CODE REVIEW

**Comment by Shrekster** [comment]
@redgoat650 @julio-lopez I have not yet gotten to writing some unit-tests around this, but would be great to get some thoughts on this now.

**Comment by Shrekster** [comment]
@redgoat650 @julio-lopez PTAL, thanks!

File: repo/open.go [hunk 1]
```diff
@@ -379,21 +387,21 @@ func readAndCacheFormatBlobBytes(ctx context.Context, st blob.Storage, cacheDire
 	if cacheEnabled {
 		b, err := readFormatBlobBytesFromCache(ctx, cachedFile, validDuration)
```
Comment (redgoat650) [inline repo/open.go:388]: Do these two functions need to be renamed? They aren't exclusively used for `formatBlob`'s anymore

File: repo/open.go [hunk 2]
```diff
@@ -362,8 +370,8 @@ func readFormatBlobBytesFromCache(ctx context.Context, cachedFile string, validD
 	return os.ReadFile(cachedFile) //nolint:wrapcheck,gosec
 }
 
-func readAndCacheFormatBlobBytes(ctx context.Context, st blob.Storage, cacheDirectory string, validDuration time.Duration) ([]byte, error) {
-	cachedFile := filepath.Join(cacheDirectory, "kopia.repository")
+func readAndCacheRepositoryBlobBytes(ctx context.Context, st blob.Storage, cacheDirectory, blobID string, validDuration time.Duration) ([]byte, error) {
+	cachedFile := filepath.Join(cacheDirectory, blobID)
 
 	if validDuration == 0 {
 		validDuration = defaultFormatBlobCacheDuration
```
Comment (redgoat650) [inline repo/open.go:377]: maybe this default value should be renamed too? Applies to more than just format blobs

File: cli/command_repository_create.go [hunk 3]
```diff
@@ -42,6 +46,8 @@ func (c *commandRepositoryCreate) setup(svc advancedAppServices, parent commandP
 	cmd.Flag("object-splitter", "The splitter to use for new objects in the repository").Default(splitter.DefaultAlgorithm).EnumVar(&c.createSplitter, splitter.SupportedAlgorithms()...)
 	cmd.Flag("create-only", "Create repository, but don't connect to it.").Short('c').BoolVar(&c.createOnly)
 	cmd.Flag("format-version", "Force a particular repository format version (1 or 2, 0==default)").IntVar(&c.createFormatVersion)
+	cmd.Flag("retention-mode", "Set the blob retention-mode for supported storage backends.").EnumVar(&c.retentionMode, minio.Governance.String(), minio.Compliance.String())
+	cmd.Flag("retention-period", "Set the blob retention-period for supported storage backends.").DurationVar(&c.retentionPeriod)
```
Comment (redgoat650) [inline cli/command_repository_create.go:50]: I assume a later PR will introduce a mechanism to change the values of an existing repo, right? I think this one only sets them on repo create?

**Review by redgoat650 — COMMENTED** [review]
Nice. A couple questions:

File: repo/initialize.go [hunk 4]
```diff
@@ -73,14 +86,27 @@ func Initialize(ctx context.Context, st blob.Storage, opt *NewRepositoryOptions,
...
 
+	if !retention.IsNull() {
+		retentionBytes, err := serializeRetentionBytes(format, retention, formatEncryptionKey)
+		if err != nil {
+			return errors.Wrap(err, "unable to encrypt retention bytes")
+		}
+
+		// Write the retention blob first so that we'll consider the repository
+		// corrupted if writing the format blob fails later.
+		if err := st.PutBlob(ctx, RetentionBlobID, gather.FromSlice(retentionBytes), blob.PutOptions{}); err != nil {
```
[the remark this answers is not in the document]
Comment (Shrekster) [reply repo/initialize.go:105]: That is a good question, same applied to the format blob, session blobs etc. Now that I think about it again, from the use-case perspective we can probably assume that all blobs that are "critical to the restore process" should have the retention settings applied.

File: repo/open.go (hunk 1 above)
Comment (Shrekster) [reply repo/open.go:388]: Yes, will rename them too.

File: repo/open.go (hunk 2 above)
Comment (Shrekster) [reply repo/open.go:377]: Yes. Will change.

File: cli/command_repository_create.go (hunk 3 above)
Comment (Shrekster) [reply cli/command_repository_create.go:50]: Correct, for now just on-create. I will add another PR after this to allow updates post create.

**Comment by Shrekster** [comment]
@julio-lopez PTAL

File: cli/command_repository_create.go (hunk 3 above)
Comment (julio-lopez) [reply cli/command_repository_create.go:50]: While this is general, should this be specific to S3 blob-stores in the mean time?

File: internal/blobtesting/map.go [hunk 5]
```diff
@@ -17,11 +17,15 @@ import (
 // DataMap is a map of blob ID to their contents.
 type DataMap map[blob.ID][]byte
 
+// versionedDataMap is a map of blob ID to a slice with versions of contents.
+type versionedDataMap map[blob.ID][][]byte
```
Comment (julio-lopez) [inline internal/blobtesting/map.go:21]: Should there be a separate `versionedMapStorage` implementation that is used only for versioned store tests? or use a single implementation (as proposed here) in all cases?

Either way, the changes related to `blobtesting/map` become a good candidate for a separate, self-contained PR. We can make those changes independently, make sure it does not break existing testing, merge upstream and iterate.

File: internal/blobtesting/map.go [hunk 6]
```diff
@@ -98,8 +105,13 @@ func (s *mapStorage) DeleteBlob(ctx context.Context, id blob.ID) error {
 	s.mutex.Lock()
 	defer s.mutex.Unlock()
 
+	if r, ok := s.retentionTime[id]; ok && r.After(s.timeNow()) {
```
Comment (julio-lopez) [inline internal/blobtesting/map.go:108]: This needs to emulate the behavior of versioned stores.
- `DeleteBlob` should add a deletion marker
- A separate `DeleteBlobVersion` should be used to delete specific version, and for that particular case then the retention period should be checked.

File: internal/blobtesting/map.go [hunk 7]
```diff
@@ -98,8 +105,13 @@ func (s *mapStorage) DeleteBlob(ctx context.Context, id blob.ID) error {
...
+		return errors.New("cannot delete object before retention period expires")
+	}
+
 	delete(s.data, id)
 	delete(s.keyTime, id)
+	delete(s.retentionTime, id)
```
Comment (julio-lopez) [inline internal/blobtesting/map.go:114]: The blob should not be deleted here, instead a deletion marker should be added.

File: internal/blobtesting/map.go (hunk 5 above)
Comment (Shrekster) [reply internal/blobtesting/map.go:21]: agreed, I tried to put something together quickly, but this really belongs is a dedicated PR

File: internal/blobtesting/map.go (hunk 7 above)
Comment (Shrekster) [reply internal/blobtesting/map.go:114]: yes, I thought about it but wasn't sure about the required guarantees around the api behavior around map storage

if this is needed then I should really take this change in a separate pr as you proposed

File: internal/epoch/epoch_manager.go [hunk 8]
```diff
@@ -674,7 +674,7 @@ func (e *Manager) GetCompleteIndexSet(ctx context.Context, maxEpoch int) ([]blob
 }
 
 // WriteIndex writes new index blob by picking the appropriate prefix based on current epoch.
-func (e *Manager) WriteIndex(ctx context.Context, dataShards map[blob.ID]blob.Bytes) ([]blob.Metadata, error) {
+func (e *Manager) WriteIndex(ctx context.Context, dataShards map[blob.ID]blob.Bytes, retentionMode string, retentionPeriod time.Duration) ([]blob.Metadata, error) {
```
Comment (julio-lopez) [inline internal/epoch/epoch_manager.go:677]: Can the retention parameters be encapsulated and retained in the blob storage provider (`e.st`), instead of having to wire these parameters through for every call?

Comment (Shrekster) [reply internal/epoch/epoch_manager.go:677]: I think we might be able to do that.

File: repo/change_password.go [hunk 9]
```diff
@@ -28,18 +32,36 @@ func (r *directRepository) ChangePassword(ctx context.Context, newPassword strin
 
 	r.formatEncryptionKey = newFormatEncryptionKey
 
-	if err := encryptFormatBytes(f, repoConfig, newFormatEncryptionKey, f.UniqueID); err != nil {
+	if err = encryptFormatBytes(f, repoConfig, newFormatEncryptionKey, f.UniqueID); err != nil {
```
Comment (julio-lopez) [inline repo/change_password.go:35]: Why is this change needed? the scope of `err` here can remain local to the block.

Comment (Shrekster) [reply repo/change_password.go:35]: I think this was due to some tests triggering on it, I'll attempt reverting.

File: internal/blobtesting/map.go (hunk 6 above)
Comment (Shrekster) [reply internal/blobtesting/map.go:108]: I do not think there is a need yet to extend the storage api with `DeleteBlobVersion` for blob-retention testing. There are a couple other APIs (SetTime/TouchBlob) etc that will suffice for now I believe. See how I am using those in unit-tests to check for retention settings.

Ideally, you'd want to verify hard-deletes (delete-by-version) as well, but that expands the scope of this PR, so could be a whole another PR after we are done with blob-retention.

File: internal/epoch/epoch_manager.go (hunk 8 above)
Comment (Shrekster) [reply internal/epoch/epoch_manager.go]: I think there is some chicken-egg problem with the compactor method which still resides within ibm-v1.

However, my latest commit moves the retention params under the epoch manager with minimal changes, PTAL

File: internal/blobtesting/map.go (hunk 6 above)
Comment (julio-lopez) [reply internal/blobtesting/map.go]: @Shrekster The blob retention settings apply specifically to `DeleteBlobVersion` and no other blob API calls.

When `versionId` is specified in an [S3 DeleteObject call](https://docs.aws.amazon.com/AmazonS3/latest/API/API_DeleteObject.html#API_DeleteObject_RequestSyntax), then the call will fail if the retention period has not elapsed.

Comment (Shrekster) [reply internal/blobtesting/map.go]: @julio-lopez Correct, but I noticed we have some extended storage APIs called `SetTime` & `TouchBlob` as well to test mutations. I'd assume they would NOT behave any differently with retention when compared to `DeleteBlobVersion`. I am obviously also making an assumption that these extensions are specifically mutating the last version and NOT pushing a new version.

I am a little sensitive to adding test-only interfaces to the APIs unless needed. Let me know if you feel strongly on whether they should push a new version too, then I can go ahead and make that change.

**Comment by Shrekster** [comment]
closing in favor of https://github.com/kopia/kopia/pull/1596
